pkg/cli/debug/debug.go

```diff
@@ -453,20 +456,40 @@ func (o *DebugOptions) RunDebug() error {
 			}
 			return fmt.Errorf(msg)
 			// switch to logging output
-		case err == krun.ErrPodCompleted, err == conditions.ErrContainerTerminated, !o.Attach.Stdin:
-			return logs.LogsOptions{
-				Object: pod,
-				Options: &corev1.PodLogOptions{
-					Container: o.Attach.ContainerName,
-					Follow:    true,
-				},
-				RESTClientGetter: o.RESTClientGetter,
-				ConsumeRequestFn: logs.DefaultConsumeRequest,
-				IOStreams:        o.IOStreams,
-				LogsForObject:    o.LogsForObject,
-			}.RunLogs()
+		case err == krun.ErrPodCompleted, err == conditions.ErrContainerTerminated:
+			o.logsOptions(pod)
+		case err == conditions.ErrNonZeroExitCode:
+			r, w := io.Pipe()
+			streams := genericclioptions.IOStreams{
+				In:     r,
+				Out:    w,
+				ErrOut: ioutil.Discard,
+			}
+			go func() {
+				fmt.Fprintf(w, "%v", logs.LogsOptions{
+					Object: pod,
+					Options: &corev1.PodLogOptions{
+						Container: o.Attach.ContainerName,
+						Follow:    true,
+					},
+					RESTClientGetter: o.RESTClientGetter,
+					ConsumeRequestFn: logs.DefaultConsumeRequest,
+					IOStreams:        streams,
+					LogsForObject:    o.LogsForObject,
+				}.RunLogs())
+				w.Close()
+			}()
+			buf := new(bytes.Buffer)
+			buf.ReadFrom(r)
+			trimmed := strings.TrimSuffix(buf.String(), "<nil>")
```

Comment:
What's this?

Reply:
the only way I found to redirect the output (non-error) to an error (which is what we want whenever container exits non-zero) have to capture the output and create the error. the output of `fmt.Fprintf(w, "%v", logs.LogsOptions... ` is 
"the logs string <nil>" the <nil> comes from the fact that the function `LogsOptions` returns an error (which is nil, even when the logs indicate the container errored out) so that is trimmed.

Reply:
@soltysh I've reworked the fn to return logs if container fails, i think it's much better now, take a look - perhaps we can kill all 4 bzs now